File: src/parser/SmsParserEngine.ts
import { TransactionParser } from './TransactionParser';
import { EvcPlusParser } from './EvcPlusParser';
import { EDahabParser } from './EDahabParser';
import { JeebParser } from './JeebParser';
import { GenericSomaliMobileMoneyParser } from './GenericSomaliMobileMoneyParser';
import { ParseResult } from '../types';

export class SmsParserEngine {
  private parsers: TransactionParser[];

  constructor() {
    this.parsers = [
      new EvcPlusParser(),
      new EDahabParser(),
      new JeebParser(),
      new GenericSomaliMobileMoneyParser(),
    ];
  }

  parseSms(sender: string, body: string): ParseResult | null {
    for (const parser of this.parsers) {
      if (parser.canParse(sender, body)) {
        const result = parser.parse(sender, body);
        if (result !== null) {
          return result;
        }
      }
    }
    return null;
  }
}

export const defaultParserEngine = new SmsParserEngine();

File: app/src/main/java/com/example/smsreaderpro/data/parser/SmsParserEngine.kt
package com.example.smsreaderpro.data.parser

import com.example.smsreaderpro.data.model.ParseResult

class SmsParserEngine(
    private val parsers: List<TransactionParser> = listOf(
        EvcPlusParser(),
        EDahabParser(),
        JeebParser(),
        GenericSomaliParser()
    )
) {
    fun parseSms(sender: String, body: String): ParseResult? {
        for (parser in parsers) {
            if (parser.canParse(sender, body)) {
                val result = parser.parse(sender, body)
                if (result != null) {
                    return result
                }
            }
        }
        return null
    }

    companion object {
        val defaultEngine = SmsParserEngine()
    }
}
